New angular templates (part 1)

emanuelschuetze opened the pull request:
Template work still in progress. This is a first "beta" state of my work for further testing.
Please review and merge!

Users app:
- Show permissions in group form (Fixed #1500)
- Added experimental csv import for users.

Agenda app:
- Agenda sort view.
- Added angular-xeditable for agenda list columns.

General:
- Remove old template and js files. Remove old views.

normanjaeckel commented on openslides/agenda/main_menu.py:
```diff
@@ -10,5 +10,5 @@ class AgendaMainMenuEntry(MainMenuEntry):
     required_permission = 'agenda.can_see'
     default_weight = 20
-    pattern_name = 'item_overview'
+    pattern_name = '/agenda'
```
Ah, that's not a pattern_name but an absolute URL. Is there a possibility of naming URL patterns in AngularJS? If not we should rename this attribute.

emanuelschuetze replied:
@normanjaeckel I used `pattern_name` only temporary (that we can use links on main menu entrys for testing) until we have fixed generic main menu with angular (#1469). I added TODOs in the main_menu.py code.

normanjaeckel commented on openslides/assignments/main_menu.py:
```diff
@@ -10,5 +10,5 @@ class AssignmentMainMenuEntry(MainMenuEntry):
     required_permission = 'assignments.can_see'
     default_weight = 40
-    pattern_name = 'assignment_list'
+    pattern_name = '/assignments'
```
Here the same.

normanjaeckel commented on openslides/core/main_menu.py:
```diff
@@ -11,4 +11,4 @@ class DashboardMainMenuEntry(MainMenuEntry):
     required_permission = 'core.can_see_dashboard'
     default_weight = 10
     icon_css_class = 'glyphicon-home'
-    pattern_name = 'core_dashboard'
```
Here the same.

normanjaeckel commented on openslides/users/main_menu.py:
```diff
@@ -10,5 +10,5 @@ class UserMainMenuEntry(MainMenuEntry):
     required_permission = 'users.can_see_extra_data'
     default_weight = 50
-    pattern_name = 'core_dashboard'
+    pattern_name = '/users'
```
Here the same.

normanjaeckel commented:
Nice. I did not check JS and HTML. This we can refactor later.

ostcar commented on bower.json:
```diff
@@ -7,9 +7,15 @@
+    "angular-ui-tree": "~2.2.0",
     "angular-gettext": "~2.0.2",
+    "angular-sanitize": "~1.3.15",
+    "angular-xeditable": "~0.1.9",
```
Have you checked the Licence of all the new dependencies? We can change the dependencies in the readme before the release, but if we can not use a dependency, we should know it now.

Also, are all this dependencies active developed?

emanuelschuetze replied:
I checked the licenses and the last release dates of the new angular modules:
 * angular-csv-import: MIT, last release 8.4.2015
 * angular-loading-bar: MIT, last release 24.2.2015
 * angular-ui-select: MIT, last release: 17.3.2015
 * angular-ui-tree: MIT, last release: 30.1.2015
 * angular-sanitize: MIT (part of angularjs)
 * angular-xeditable: MIT, last release: 26.3.2015

Conclusion: Yes, all packages are Free Software and active developed.

ostcar commented on bower.json:
```diff
@@ -7,9 +7,15 @@
     "bootstrap": "~3.3.1",
     "datatables-bootstrap3-plugin": "~0.2.0",
-    "angular": "~1.3.13",
+    "angular": "~1.3.15",
```
Is there a reason why you changed the dependency for angular? Is 1.3.13 not supported anymore?

emanuelschuetze replied:
At least one packages has a dependency to angular 1.3.15. So I updated angular to last stable release.

ostcar commented on openslides/agenda/static/templates/agenda/item-csv-import.html:
```diff
@@ -0,0 +1,39 @@
+
+<p translate>Select a CSV file to import agenda items!
+
+<p translate>Please note:</p>
```
The closing p-tag can be removed

emanuelschuetze replied:
No, not in this case. We need the </p> for closing the translation string.

ostcar replied:
Have you tested this? If the closing p-tag is really required, then you have to add one in line 10

ostcar commented on openslides/agenda/static/templates/agenda/item-csv-import.html:
```diff
@@ -0,0 +1,39 @@
+  <div class="form-group">
+    <label for="InputFile" translate>CSV file:</label>
+    <input type="file" id="InputFile">
+    <p class="help-block" translate>The file has to be encoded in UTF-8.
```
Is this still required with angular-csv?

emanuelschuetze replied:
currently utf8 is still required so that we can print the csv output in right encoding in the template (which is utf8 encoded). But I added a TODO to check this later.

ostcar replied:
Where is the TODO?

ostcar commented on openslides/agenda/static/templates/agenda/item-form.html:
```diff
@@ -1,8 +1,56 @@
+  </div>
+  <div class="form-group">
+    <label for="selectParent" translate>Parent</label>
+    <select ng-options="item.id as item.get_title for item in items" ng-model="item.parent" class="form-control" name="selectParent">
```
Please make this line shorter

emanuelschuetze replied:
done

ostcar commented on openslides/agenda/static/templates/agenda/item-list.html:
```diff
@@ -1,9 +1,94 @@
+      <td>{{ item.item_number }}
+      <td><a ui-sref="agenda.item.detail({id: item.id})">{{ item.title }}</a>
+      <td os-perms="agenda.can_manage" class="optional">
+        <a href="#" editable-number="item.duration" e-min="1" onaftersave="save(item)">{{ item.duration || '%' }}</a>
```
Please make this line shorter

emanuelschuetze replied:
done

ostcar commented on openslides/agenda/static/templates/agenda/item-sort.html:
```diff
@@ -0,0 +1,26 @@
+  <ol ui-tree-nodes="" ng-model="items">
+    <li ng-repeat="item in items" ui-tree-node>
+      <div ui-tree-handle>
+        ({{item.weight}}) {{item.title}}
```
Please add spaces after/before {{ and }}

emanuelschuetze replied:
done

ostcar commented on openslides/agenda/static/templates/agenda/item-sort.html:
```diff
@@ -0,0 +1,26 @@
+          <div ui-tree-handle>
+            # {{subItem.title}}
+          </div>
+        </li>
```
Can be removed:
```An li element's end tag may be omitted if the li element is immediately followed by another li element or if there is no more content in the parent element.```

emanuelschuetze replied:
done

ostcar commented on openslides/agenda/static/templates/agenda/item-sort.html:
```diff
@@ -0,0 +1,26 @@
+          </div>
+        </li>
+      </ol>
+    </li>
```
Can be removed

emanuelschuetze replied:
done

ostcar commented on openslides/agenda/urls.py:
```diff
@@ -4,97 +4,14 @@
+    # PDF
     url(r'^print/$',
         views.AgendaPDF.as_view(),
         name='print_agenda'),
```
Please rename the name to something that starts with "agenda"

emanuelschuetze replied:
done

ostcar commented on openslides/assignments/static/templates/assignments/assignment-detail.html:
```diff
@@ -1,2 +1,31 @@
+    <i class="fa fa-angle-double-left fa-lg"></i>
+    <translate>Back to overview</translate>
+  </a>
+  <a href="/assignments/{{assignment.id}}/print/" target="_blank" class="btn btn-default btn-sm">
```
Plase use the url-pattern-name

emanuelschuetze replied:
done

ostcar commented on openslides/assignments/static/templates/assignments/assignment-list.html:
```diff
@@ -1,7 +1,72 @@
+    <i class="fa fa-plus fa-lg"></i>
+    <translate>New</translate>
+  </a>
+  <a href="/assignments/print/" target="_blank" class="btn btn-default btn-sm">
```
Please use url-pattern-name

emanuelschuetze replied:
done

ostcar commented on openslides/urls.py:
```diff
@@ -12,21 +12,20 @@
+    url(r'^assignments/', include('openslides.assignments.urls')),
+    url(r'^assignments.*', IndexView.as_view()),
+    url(r'^agenda/', include('openslides.agenda.urls')),
+    url(r'^agenda.*', IndexView.as_view()),
```
Please add a TODO that all patterns with end with ".*" can be removed after we add the global url-pattern "/" to the indexView

emanuelschuetze replied:
done

ostcar commented on openslides/agenda/static/templates/agenda/item-list.html:
```diff
@@ -1,9 +1,94 @@
+    <i class="fa fa-download fa-lg"></i>
+    <translate>Import</translate>
+  </a>
+  <a href="/agenda/print/" target="_blank" class="btn btn-default btn-sm">
```
Please use the url-pattern-name

emanuelschuetze replied:
done

ostcar commented on openslides/core/templates/base.html:
```diff
@@ -89,13 +89,14 @@
-                        <a href="{{ entry.get_url }}" class="tooltip-right">
-                        <!--TODO-->
+                    <li>
+                        <a href="{{ entry.pattern_name }}" class="tooltip-right">
```
Please use ui-sref

emanuelschuetze replied:
this base template is not an angular template and will be removed is the template work finished. My changes are only for getting faster the specific app url for testing. So I don't want to do more effort in this depricated template.

ostcar replied:
ok

emanuelschuetze commented:
all done. ready for last review.

Note: There are a lot of tests which failed after removing the unused django views. What should we do with these tests now? Removing all failed tests?

normanjaeckel commented:
Travis fails not because of old tests I think? @emanuelschuetze please check.

emanuelschuetze commented:
@normanjaeckel Now travis failed with the old(!) tests.

emanuelschuetze commented:
Added more commits for adding angular template for motions/categroy/tags/medieafiles.

emanuelschuetze commented:
Skip old tests to fix travis. Ready for review.

emanuelschuetze commented:
* motions/urls.py: renamed url_patterns
* motions/views.py: all unused views removed.
* assignments/urls.py: renamed url_patterns
* mediefiles/urls.py ../views.py: this pdf views will be changed after we find a solution to update the file presentation widget. The other views are removed.

emanuelschuetze commented:
@ostcar all done

emanuelschuetze commented:
Rebased on master. Ready for review.

Still open: use new item tree, use new assignment candidate functionality

normanjaeckel commented:
Ok.

emanuelschuetze commented:
Fixed agenda sort view and item view by using new item tree via rest api.
@ostcar ready for last review!

ostcar commented on openslides/agenda/static/js/agenda/agenda.js:
```diff
@@ -40,30 +62,105 @@ angular.module('OpenSlidesApp.agenda', [])
-    $scope.test_singular = i18n.ngettext('test', 'tests', 1);
+
+    // get full agenda tree
+    $http.get('/rest/agenda/item/tree')
```
Please put this line inside the resolve-object in line 17 (I added also a comment to your commit with a description how to do this)

ostcar commented on openslides/agenda/static/js/agenda/agenda.js:
```diff
@@ -40,30 +62,105 @@ angular.module('OpenSlidesApp.agenda', [])
+.controller('ItemCreateCtrl', function($scope, $state, $http, Agenda) {
+    Agenda.bindAll({}, $scope, 'items');
+    // get all item types via OPTIONS request
+    $http({ 'method': 'OPTIONS', 'url': '/rest/agenda/item/' })
```
Please call the request in the resolve-object.

ostcar commented on openslides/agenda/static/js/agenda/agenda.js:
```diff
@@ -40,30 +62,105 @@ angular.module('OpenSlidesApp.agenda', [])
+.controller('ItemUpdateCtrl', function($scope, $state, $http, Agenda, item) {
+    Agenda.bindAll({}, $scope, 'items');
+    // get all item types via OPTIONS request
+    $http({ 'method': 'OPTIONS', 'url': '/rest/agenda/item/' })
```
Here the same

ostcar commented on openslides/agenda/static/js/agenda/agenda.js:
```diff
@@ -40,30 +62,105 @@ angular.module('OpenSlidesApp.agenda', [])
+.controller('AgendaSortCtrl', function($scope, $http, Agenda) {
+    Agenda.bindAll({}, $scope, 'items');
+    // get full agenda tree
+    $http.get('/rest/agenda/item/tree')
```
Here the same

ostcar commented on openslides/assignments/static/js/assignments/assignments.js:
```diff
@@ -41,29 +41,61 @@ angular.module('OpenSlidesApp.assignments', [])
+.controller('AssignmentListCtrl', function($scope, $http, Assignment) {
     Assignment.bindAll({}, $scope, 'assignments');
+    // get all item types via OPTIONS request
+    $http({ 'method': 'OPTIONS', 'url': '/rest/assignments/assignment/' })
```
Plase do this in the resolve-object in the stateProvider

ostcar commented on openslides/mediafiles/urls.py:
```diff
@@ -4,21 +4,6 @@
-    url(r'^(?P<pk>\d+)/del/$',
-        views.MediafileDeleteView.as_view(),
-        name='mediafile_delete'),
     url(r'^pdf/next/$', views.PdfNextView.as_view(), name='next_pdf_page'),
```
Could you change the name of the following url-patterns to "mediafiles_*"? I know, this is not part of you pull request, but then we can close #1509

ostcar commented on openslides/users/static/js/users/users.js:
```diff
@@ -300,19 +332,28 @@ angular.module('OpenSlidesApp.users', [])
-    //TODO: permissions Group.bindAll({}, $scope, 'groups');
+.controller('GroupCreateCtrl', function($scope, $state, $http, Group) {
+    // get all permissions via OPTIONS request
+    $http({ 'method': 'OPTIONS', 'url': '/rest/users/group/' })
```
Please to this in the resolve-object

ostcar commented on openslides/users/static/js/users/users.js:
```diff
@@ -300,19 +332,28 @@ angular.module('OpenSlidesApp.users', [])
-.controller('GroupUpdateCtrl', function($scope, $state, Group, group) {
+.controller('GroupUpdateCtrl', function($scope, $state, $http, Group, group) {
+    // get all permissions via OPTIONS request
+    $http({ 'method': 'OPTIONS', 'url': '/rest/users/group/' })
```
Please do this thins in the resolve-object

ostcar commented on tests/old/account/test_widgets.py:
```diff
@@ -51,6 +53,7 @@ def test_widget_appearance(self):
         response = self.client.get('/dashboard/')
         self.assertContains(response, 'My personal info', status_code=200)
 
+    @skip
```
You could have skipped a hole test-class by putting the decorator in front of the class. But your way works, too.

ostcar commented:
Looks really good! 

I found some strange behaviour in the agenda app
* In the list-view, I can not click on the title of an item to get to the detail-view (this is the only way for normal users to see the item)
* The pdf-link in the list-view does not work when not logged in
*  The filter in the list view seems not to work.
* In the change sort order view, there is not button to save the changed tree ( but it seems the tree is saved automatically maybe there only need to be a text to explain this)
* I can not see in the list view, when an item is a child-item. Maybe there should be more spaces before the title
* I am confused about the '%' symbol in duration. You probably mean that no time is set. I don't get this from the sign
* There should be an information that the duration is in minutes
* The delete button does not work
* In the create view it is not clear which fields are required
* In the create formular, if you leave any field blank, no item is created, but you are redirected to the overview.

emanuelschuetze commented:
@ostcar I fixed a lot of your reported issues. Some points are opened as new issues. All done. Ready for next review!

Note: In my last commit is still a TODO because I get a strange error if ng-fab-forms is includes (which does not happens with a clean ng-fab-forms demo; so maybe a OpenSides/angular problem?) Any idea?

emanuelschuetze commented:
I just added some more template improvements (ui-bootstrap and ngbootbox). Please let me know when I can rebase all commits.

ostcar commented:
Looks good. I would say that we merge all this and fix any problem later.

So please rebase

emanuelschuetze commented:
Rebased.

emanuelschuetze commented:
Rebased again after merging #1514 .

ostcar commented:
@normanjaeckel Do you want to look at this? If not, please merge

emanuelschuetze commented:
@normanjaeckel ping
